source/fuzz/fuzzer_pass_donate_modules.cpp

```diff
@@ -493,9 +575,79 @@ void FuzzerPassDonateModules::HandleFunctions(
     // a fresh id.  This is necessary because functions include forward
     // references, e.g. to labels.
-    function_to_donate->ForEachInst([this, &original_id_to_donated_id](
+    function_to_donate->ForEachInst([this, donor_ir_context,
+                                     &original_id_to_donated_id](
                                         const opt::Instruction* instruction) {
-      if (instruction->result_id()) {
+      if (!instruction->result_id()) {
+        return;
+      }
+      if (IgnoreInstruction(instruction)) {
+        if (instruction->opcode() == SpvOpArrayLength) {
+          // We treat the OpArrayLength instruction specially.  In the donor
+          // shader this gets the length of a runtime array that is the final
+          // member of a struct.  During donation, we will have converted the
+          // runtime array type, and the associated struct field, into a fixed-
+          // size array.  We can then use the constant size of this fixed-sized
+          // array wherever we would have used the result of an OpArrayLength
+          // instruction.
+          uint32_t donated_variable_id = original_id_to_donated_id->at(
+              instruction->GetSingleWordInOperand(0));
+          auto donated_variable_instruction =
+              GetIRContext()->get_def_use_mgr()->GetDef(donated_variable_id);
+          auto pointer_to_struct_instruction =
+              GetIRContext()->get_def_use_mgr()->GetDef(
+                  donated_variable_instruction->type_id());
+          assert(pointer_to_struct_instruction->opcode() == SpvOpTypePointer &&
+                 "Type of variable must be pointer.");
+          auto donated_struct_type_instruction =
+              GetIRContext()->get_def_use_mgr()->GetDef(
+                  pointer_to_struct_instruction->GetSingleWordInOperand(1));
+          assert(
+              donated_struct_type_instruction->opcode() == SpvOpTypeStruct &&
+              "Pointee type of pointer used by OpArrayLength must be struct.");
+          assert(donated_struct_type_instruction->NumInOperands() ==
+                     instruction->GetSingleWordInOperand(1) + 1 &&
+                 "OpArrayLength must refer to the final member of the given "
+                 "struct.");
+          uint32_t fixed_size_array_type_id =
+              donated_struct_type_instruction->GetSingleWordInOperand(
+                  donated_struct_type_instruction->NumInOperands() - 1);
+          auto fixed_size_array_type_instruction =
+              GetIRContext()->get_def_use_mgr()->GetDef(
+                  fixed_size_array_type_id);
+          assert(fixed_size_array_type_instruction->opcode() ==
+                     SpvOpTypeArray &&
+                 "The donated array type must be fixed-size.");
+          original_id_to_donated_id->insert(
+              {instruction->result_id(),
+               fixed_size_array_type_instruction->GetSingleWordInOperand(1)});
```

Comment:
Could you may add yet another local variable here to clarify that this is the id corresponding to the array size?

Reply:
Done (though I hate local variables :-))